File: JAVA/PartitionProblemMemoization/src/partitionproblemmemoization/Memoization.java
package partitionproblemmemoization;

public class Memoization {
    private final int[] num;
    private Boolean[][] map;
    private final int n;
    private int sum = 0;
    private boolean found;
    
    public Memoization(int[] num) {
        this.num = num;
        this.n = num.length;
    }
    
    public void printVector(){
        System.out.print("| ");
        for (int i = 0; i < n; i++) { 
            System.out.print(num[i] + " | "); 
        } 
    }
    
    private void getResult(){
        System.out.println("Results: " + sum);
        
        String set1 = "{";
        String set2 = "{";
        
        int x = n; 
        int currSum = sum; 
      
        while (x > 0 && currSum >= 0){
            
            if (map[x-1][currSum]){ 
                x--;
                set1 += num[x] + ", ";                
            }else if (map[x-1][currSum-num[x-1]]){ 
                x--; 
                currSum -= num[x]; 
                set2 += num[x] + ", "; 
            } 
        } 
        
        set1 = set1.substring(0, set1.length()-2) + '}';
        set2 = set2.substring(0, set2.length()-2) + '}';
        System.out.println("Set 1: " + set1);
        System.out.println("Set 2: " + set2);
        
    }
    
    public void printResult(){
        if(found){
            System.out.println("\nVector can be partitioned");       
            getResult();
        }else{
            System.out.println("\nVector can not be partitioned");
        }
    }
    
    public void printTime(long timeMillis, long timeNano, String scale){
        System.out.print("For " + n + " elements -> " + "Taken time: ");
        float print = (float) ((timeMillis == 0)? timeNano/1e9d : timeMillis/1e3d);
        if(scale == null){            
            System.out.printf("%.8f s.\n", print);
        }else{
            switch (scale) {
                case "ns": System.out.println(timeNano + " " + scale + '.');
                break;
                case "ms": System.out.println(timeMillis + " " + scale + '.');
                break;
                default: System.out.printf("%.8f s.\n", print);
                break;
            }  
        }
    }
    
    private int add(){
        for (int i = 0; i < n; i++) sum += num[i];        
        return sum;
    }
    
    public boolean canPartition() {
        int su = add();        
        return (su % 2 == 0);
    }
    
    public void canPartitionRecursiveAux(){
        sum /= 2;  
        map = new Boolean[n+1][sum+1];
        for(int i=0; i <= n; i++) map[i][0] = true;

        for(int s=1; s <= sum ; s++) map[0][s] = false;
        found = canPartitionRecursive(n, sum);
    }

    private boolean canPartitionRecursive(int i, int s) {
        if(map[i][s] != null){
            return map[i][s];
        }
        
        if (s <= 0) return true;
        
        if (i == 0) return false;
        
        boolean result;
        if(s < num[i-1]){
            result = canPartitionRecursive(i-1, s);
        }else{
            result = canPartitionRecursive(i-1, s) || 
                    canPartitionRecursive(i-1, s-num[i-1]);
        }
        
        map[i][s] = result;
        
        return result;
    }
}

File: CPP/PartitionProblemMemoization/Memoization.cpp
#include "Memoization.h"

Memoization::Memoization(int* num, int size) {
    this->num = num;
    this->n = size;       
}

void Memoization::printVector() {
    cout << "| ";
    for (int i = 0; i < n; i++) {
        cout << num[i] << " | ";
    }
    cout << endl;
}

void Memoization::getResult() {
    cout << "Results: " << sum << endl;
    
    string set1 = "{";
    string set2 = "{";
    int x = n;
    int curr = sum;
    
    while (x > 0 && curr >= 0){
            
            if (mp[x-1][curr] == '1'){ 
                x--;
                set1 += to_string(num[x]) + ", ";                
            }else if (mp[x-1][curr-num[x-1]] == '1'){ 
                x--; 
                curr -= num[x]; 
                set2 += to_string(num[x]) + ", "; 
            } 
        }
    
    set1 = set1.substr(0, set1.length()-2) + '}';
    set2 = set2.substr(0, set2.length()-2) + '}';
    cout << "Set 1: " << set1 << endl;
    cout << "Set 2: " << set2 << endl;
}

void Memoization::printResult() {
    if(found == '1'){
        cout << "\nVector can be partitioned" << endl;
        Memoization::getResult();
    }else{
        cout << "\nVector can not be partitioned" << endl;
    }
}

void Memoization::printTime(long timeMS, long timeNS, string scale) {
    printf("For %d elements -> Taken time: ", n); 
    float print = (float) ((timeMS == 0)? timeNS/1e9 : timeMS/1e3);
    if(scale.empty()){
        printf("%.8f s.\n", print);
    }else{
        if(scale == "ns"){
            cout << timeNS << " " << scale << '.' << endl;
        }else if(scale == "ms"){
            cout << timeMS << " " << scale << '.' << endl;
        }else{
            printf("%.8f s.\n", print);
        }
    }
}

int Memoization::add() {
    for (int i = 0; i < n; i++) sum += num[i];
    return sum;
}

bool Memoization::canPartition() {
    int sum = add();
    return (sum % 2 == 0);
}

void Memoization::canPartitionAux() {
    sum /= 2; 
    mp = new char*[n+1];    
    
    for(int a = 0; a <= n; a++) mp[a] = new char[sum+1];
    
    for(int i=0; i <= n; i++) mp[i][0] = '1';

    for(int s=1; s <= sum ; s++) mp[0][s] = '0';    
    
    found = Memoization::canPartitionMemo(n, sum);
}

char Memoization::canPartitionMemo(int i, int s) {
    if(mp[i][s] != '\0'){
        return mp[i][s];
    }
    
    if(s <= 0) return '1';
    
    if(i == 0) return '0';
    
    char result;
    if(s < num[i-1]){
        result = Memoization::canPartitionMemo(i-1, s);
    }else{
        result = Memoization::canPartitionMemo(i-1, s) |
                Memoization::canPartitionMemo(i-1, s-num[i-1]);
    }
    
    mp[i][s] = result;
    
    return result;
}




